src/python/pants/backend/javascript/dependency_inference/rules.py

```diff
@@ -101,23 +121,107 @@ async def map_candidate_node_packages(
     )
 
 
-@rule
-async def prepare_inference_metadata(imports: PackageJsonImports) -> InferenceMetadata:
+def _create_inference_metadata(
+    imports: PackageJsonImports, config: TSConfig | None
+) -> InferenceMetadata:
     return InferenceMetadata.javascript(
         imports.root_dir,
-        {pattern: list(replacements) for pattern, replacements in imports.imports.items()},
+        dict(imports.imports),
+        config.resolution_root_dir if config else None,
+        dict(config.paths or {}) if config else {},
     )
 
 
-async def _prepare_inference_metadata(address: Address) -> InferenceMetadata:
-    owning_pkg = await Get(OwningNodePackage, OwningNodePackageRequest(address))
+async def _prepare_inference_metadata(address: Address, file_path: str) -> InferenceMetadata:
+    owning_pkg, maybe_config = await concurrently(
+        find_owning_package(OwningNodePackageRequest(address)),
+        find_parent_ts_config(ParentTSConfigRequest(file_path, "jsconfig.json"), **implicitly()),
+    )
     if not owning_pkg.target:
-        return InferenceMetadata.javascript(address.spec_path, {})
-    return await Get(
-        InferenceMetadata, PackageJsonSourceField, owning_pkg.target[PackageJsonSourceField]
+        return InferenceMetadata.javascript(
+            (
+                os.path.dirname(maybe_config.ts_config.path)
+                if maybe_config.ts_config
+                else address.spec_path
+            ),
+            {},
+            maybe_config.ts_config.resolution_root_dir if maybe_config.ts_config else None,
+            dict(maybe_config.ts_config.paths or {}) if maybe_config.ts_config else {},
+        )
+    return _create_inference_metadata(
+        await subpath_imports_for_source(owning_pkg.target[PackageJsonSourceField]),
+        maybe_config.ts_config,
     )
 
 
+def _add_extensions(file_imports: frozenset[str]) -> PathGlobs:
+    extensions = JS_FILE_EXTENSIONS + tuple(f"/index{ext}" for ext in JS_FILE_EXTENSIONS)
+    return PathGlobs(
+        string
+        for file_import in file_imports
+        for string in (
+            [file_import]
+            if PurePath(file_import).suffix
+            else [f"{file_import}{ext}" for ext in extensions]
+        )
+    )
+
+
+async def _determine_import_from_candidates(
+    candidates: ParsedJavascriptDependencyCandidate,
+    package_candidate_map: NodePackageCandidateMap,
+) -> Addresses:
+    paths = await path_globs_to_paths(_add_extensions(candidates.file_imports))
+    local_owners = await Get(Owners, OwnersRequest(paths.files))
+
+    owning_targets = await Get(Targets, Addresses(local_owners))
```

Comment:
Are these some of the rules that are causing recursion grief? Or is this a result of the migration not working well inside of a rule helper?

Reply:
These are the results of moving code around instead of changing the invocation. I haven't ran the conversion tool in this PR, I did any call-by-name changes by hand.

Reply:
The recursion failure is here https://github.com/pantsbuild/pants/pull/21176/commits/4f799628d13bb32c25c14d93f0c5ef6f269ff8fd